### ISSUE TITLE
Syncing multiple daemons

### ISSUE THREAD

**Issue by alphapapa** [andrewrk/groovebasin#423]
I just discovered groovebasin, and I really like it.  I was having problems with MPD not handling replaygain correctly on some tracks, clipping and distorting them really badly.  I think it was probably a bug that could be fixed, but GB handles it fine, so I'm giving GB a try as a drop-in replacement.  I like that I can use mpc and Cantata and my mobile MPD clients.  :)

I have MPD running in two different rooms, and sometimes I use [mpdsync](http://github.com/alphapapa/mpdsync/tree/python-mpd2) to sync them.  One of the things I've been fiddling with is trying to keep them playing in sync so that there's no echo effect.  With MPD this is hit-and-miss at best: some tracks seek consistently and quickly, while others seem to seek with varying latencies.  Also, MPD's buffers are big and its latency is high by design.  And it also doesn't seem to report the elapsed track time very consistently.  So the result is that it's very hard to sync two MPD daemons, and it seems impossible to do so consistently.

Since GB can be a drop-in replacement for MPD, I tried it with my mpdsync script.  At first I noticed that the reported elapsed time was much more consistent, within a few tens of ms every time, instead of jittering by a few hundred ms.  But then I noticed that sometimes GB seemed to have strange network latency: sometimes it would only take a few ms to reply, but other times it would take a few hundred ms.  I admit I can't be certain yet whether it's GB causing that or something else on the machine, but I haven't noticed this issue with MPD.

Anyway, the whole process of seeking a slave daemon to a master over and over again until they nearly match, and doing so every time the track changes, is not a very clean way to do it, resulting in a lot of audible pops and interruptions in the slaves' audio output.

So I was wondering if you would be interested in considering this as a future feature, to sync multiple daemons over the network.  The idea, of course, is that each daemon has its own local, identical copy of the music library.  (I guess it doesn't matter whether it's a local copy or a mounted network share, but I use local copies on mine.)  I've tried HTTP streaming with MPD (not yet with GB), but it doesn't work well because of the latency.

There's an interesting project and article here: https://snarfed.org/synchronizing_mp3_playback  It concludes that to really do it right, you need to have the slaves sync their clocks with the master over the network (NTP-style I guess), and then begin playing a track at the same moment in time, adjusted for local latency.

Maybe that could also be done for seeking within a track (it would seem necessary to support seeking), but sometimes I have wondered if one of the problems is with seeking itself.  I don't know how seeking works on the decoder level, but I have wondered if seeking tends to seek to the beginning  of frames, and as a result is imprecise at a low level (which wouldn't be noticed except when trying to seek very precisely), and also perhaps seeks with unpredictable latency.  If so, maybe GB could use this information when seeking slave daemons, seeking to frames instead of a certain time reference.

Anyway, thanks for groovebasin, and thanks for any thoughts on this idea.

**Comment by andrewrk** [comment]
Are you sure this isn't best handled with third party software? Perhaps something like PulseAudio? Here's something I found with a quick search: http://www.hackerposse.com/~rozzin/journal/whole-home-pulseaudio.html

**Comment by alphapapa** [comment]
Thanks, I actually read that page a few years ago.  At the time I gave it a shot, but the PulseAudio streaming was completely unreliable.  Perhaps it's improved since then, although I don't know anything about the latency-compensating code; I have my doubts about it being sophisticated enough to sync well over a network, but who knows.  :)

But that's not the solution I'm looking for anyway, because I don't want to stream audio over my LAN.  For one thing, my wifi network has fairly poor performance, especially between upstairs/downstairs, living in an area with many other access points nearby.  But I also want to be able to play music independently at each system; I don't want to simply, in effect, have wireless speakers.  :)

I understand if you're not interested in this.  But since you're sort-of reinventing MPD from scratch, and since you know how the low-level audio code works, I thought maybe it would be easier for Groove Basin to get some support for syncing other instances than it would be to retrofit it into MPD (or kludge it with a separate script like I've been trying).

I wonder if doing something simple like this would work fairly well:

1.  Measure average latency between instances
2.  Choose a frame in the track to seek to, one that's closest to the desired seek position
3.  Send messages to other instances telling them to seek to the chosen frame in 100ms or so
4.  Set a timer for 100ms
5.  Seek to the chosen frame and start playing

Or for the slaves, steps 4-5.

I guess you could get more sophisticated and adjust based on buffer sizes, measured or reported driver latency, etc.  But this might be a good-enough solution.

Anyway, if you're not interested, that's just fine.  But maybe keep it in the back of your mind when you're working on related parts of the code... ;)

Thanks.

**Comment by andrewrk** [comment]
I don't think that syncing multiple instances of Groove Basin with library copies makes sense for the project. However what might make sense for the project is a feature where, using the Groove Basin protocol, you could synchronize playback to a specific time. For example, issue a play command where the playback should start at precisely this time, to the nanosecond. Then Groove Basin could use its knowledge of internal buffers, timing, latency, etc, to make the audio playback happen at that time.

If you wanted to accomplish your goal, using this feature, you would then need to write code which ensured that each instance had the same library and play queue at all times, make sure each server had high precision accuracy on the system clock, and then issue time-synchronized play commands.

**Comment by alphapapa** [comment]
Yes, that's what I meant, to sync the playback precisely.  Syncing the libraries is simple (unison, etc), and syncing playlists is also solved (using mpdsync).
